File: src/main/java/ru/siksmfp/rx/play/handler/impl/nio/ReadHandler.java
package ru.siksmfp.rx.play.handler.impl.nio;

import ru.siksmfp.rx.play.handler.api.Handler;
import ru.siksmfp.rx.play.utils.TransmogrifyUtils;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.SelectionKey;
import java.nio.channels.SocketChannel;
import java.util.Map;
import java.util.Queue;

import static java.nio.channels.SelectionKey.OP_WRITE;
import static ru.siksmfp.rx.play.utils.TransmogrifyUtils.transmogrify;

public class ReadHandler implements Handler<SelectionKey> {

    private Map<SocketChannel, Queue<ByteBuffer>> pendingData;

    public ReadHandler(Map<SocketChannel, Queue<ByteBuffer>> pendingData) {
        this.pendingData = pendingData;
    }

    @Override
    public void handle(SelectionKey selectionKey) throws IOException {
        SocketChannel sc = (SocketChannel) selectionKey.channel();
        ByteBuffer bb = ByteBuffer.allocateDirect(80);
        int read = sc.read(bb);

        if (read == -1) {
            pendingData.remove(sc);
            sc.close();
            System.out.println("Disconnected from in read" + sc);
            return;
        }

        if (read > 0) {
            transmogrify(bb);
            pendingData.get(sc).add(bb);
            selectionKey.interestOps(OP_WRITE);
        }
    }
}

File: src/main/kotlin/ru/siksmfp/network/harness/implementation/nio/server/ReadHandler.kt
package ru.siksmfp.network.harness.implementation.nio.server

import ru.siksmfp.network.harness.api.Handler
import ru.siksmfp.network.harness.implementation.nio.byteBufferToString
import java.nio.ByteBuffer
import java.nio.channels.SelectionKey
import java.nio.channels.SelectionKey.OP_WRITE
import java.nio.channels.SocketChannel
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class ReadHandler(
        private val clients: MutableSet<SocketChannel>,
        private val selectorActions: Queue<Runnable>
) : SelectionHandler {

    private var handler: Handler<String>? = null

    private var executorService: ExecutorService = Executors.newFixedThreadPool(1)

    override fun handle(selectionKey: SelectionKey) {
        val sc = selectionKey.channel() as SocketChannel
        val bb = ByteBuffer.allocateDirect(10000)//todo improve memory management
        val read = sc.read(bb)
        if (read == -1) {
            clients.remove(sc)
            sc.close()
            println("Disconnected from in read $sc")
            return
        }
        val response = byteBufferToString(bb, read)
        handler?.handle(response)
        println("NioServer: received $response")

        if (read > 0) {
            executorService.submit {
                selectorActions.add(Runnable {
                    selectionKey.interestOps(OP_WRITE)
                })
                selectionKey.selector().wakeup()
            }
        }
    }

    override fun close() {
        handler?.close()
        executorService.shutdown()
    }

    fun setHandler(handler: Handler<String>) {
        this.handler = handler
    }
}